sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -0,0 +1,358 @@
+    desired_bundle_count = desired_bundle_size // self.avg_doc_size
+
+    bundle_start = start_position
+    while bundle_start < stop_position:
+      bundle_end = min(stop_position, bundle_start + desired_bundle_count)
+      yield iobase.SourceBundle(weight=bundle_end - bundle_start,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=bundle_end)
+      bundle_start = bundle_end
+
+  def get_range_tracker(self, start_position, stop_position):
+    if start_position is None:
+      start_position = 0
+    if stop_position is None:
+      stop_position = self.doc_count
+    return OffsetRangeTracker(start_position, stop_position)
+
+  def read(self, range_tracker):
+    with MongoClient(self.uri, **self.spec) as client:
+      # docs is a MongoDB Cursor
+      docs = client[self.db][self.coll].find(
+          filter=self.filter, projection=self.projection
+      )[range_tracker.start_position():range_tracker.stop_position()]
+      for index in range(range_tracker.start_position(),
+                         range_tracker.stop_position()):
+        if not range_tracker.try_claim(index):
+          return
+        yield docs[index - range_tracker.start_position()]
+
+  def display_data(self):
+    res = super(_BoundedMongoSource, self).display_data()
+    res['uri'] = self.uri
+    res['database'] = self.db
+    res['collection'] = self.coll
+    res['filter'] = self.filter
+    res['project'] = self.projection
+    res['mongo_client_spec'] = self.spec
+    return res
+
+  def _get_avg_document_size(self):
+    with MongoClient(self.uri, **self.spec) as client:
+      return client[self.db].command('collstats', self.coll)['avgObjSize']
+
+  def _get_document_count(self):
+    with MongoClient(self.uri, **self.spec) as client:
+      return client[self.db][self.coll].count_documents(self.filter)
+
+
+@experimental()
+class WriteToMongoDB(PTransform):
+  """WriteToMongoDB is a ``PTransform`` that writes a ``PCollection`` of
+  mongodb document to the configured MongoDB server.
+
+  In order to make the document writes idempotent so that the bundles are
+  retry-able without creating duplicates, the PTransform added 2 transformations
+  before final write stage:
+  a ``GenerateId`` transform and a ``Reshuffle`` transform.::
+
+                  -----------------------------------------------
+    Pipeline -->  |GenerateId --> Reshuffle --> WriteToMongoSink|
+                  -----------------------------------------------
+                                  (WriteToMongoDB)
+
+  The ``GenerateId`` transform adds a random *_id* field to the documents if
+  they don't already have one, it uses the same format as MongoDB default. The
+  ``Reshuffle`` transform makes sure that no fusion happens between
+  ``GenerateId`` and the final write stage transform, so that when the
```

Comment:
, so that the set of documents and their unique IDs are not regenerated if final write step is retried due to a failure. This prevents duplicate writes of the same document with different unique IDs.